### PR TITLE
Discrete Flow Matching

### PR DESCRIPTION
**Pull request by bicycle1885**
This PR adds Discrete Flow Matching as proposed by Itai Gat, 2024, https://arxiv.org/abs/2407.15595, along with some tests and an example notebook.

### CODE REVIEW

File: src/flows.jl [hunk 1]
```diff
@@ -211,33 +232,61 @@ end
...
+end
+
+function takestep(rng, f::DiscreteFlow, xt, out, t, step, tracker)
+    κ(t) = f.schedule(t)
+    κ̇(t) = derivative(κ, t)
+    # track the current state and the predicted logits (should track probs instead?)
+    track!(tracker, t, xt, MatrixFlowState(out, xt.mask))
+    # forward velocity u_t(⋅, Xt) (equation 24)
+    velo = (κ̇(t) / (1 - κ(t))) .* (softmax(out) - xt.x)
+    p = xt + step * velo
```
Comment (murrellb) [inline src/flows.jl:265]: Please try with:
`p = clamp.(xt + step * velo, 0, Inf)`
and see if this removes the background? Hypothesis: Some fraction of the time `p` might go negative, and then your normalization will interact weirdly with `randcat`.
